third_party_licenses/README.md

```diff
@@ -0,0 +1,100 @@
+    We failed to resolve the following dependencies:
+    gomodules.xyz/jsonpatch/v2
+    honnef.co/go/tools
+    ml_metadata
+    ```
+
+3. Create a CSV for unknown repositories.
+
+    As you can see, there are 3 dependencies that can not be resolved automatically from the previous step. We have to search for these repositories manually and add a CSV file `dep_repo.manual.csv` for future runs.
+
+    Rerun the command in step 2 to resolve all repositories this time:
+    ```
+    $ python <kfp_local_repo>/third_party/cli/get_github_repo.py
+
+    ......
+    Successfully resolved github repo for 92 dependencies and saved to repo.txt. Failed to resolve 0 dependencies.
+    ```
+
+## Get all license URLs and types
+
+1. Get all license information by this command:
+    ```
+    $ python <kfp_local_repo>/third_party/cli/get_github_license_info.py --github-api-token-file=<github_token_file>
+    ......
+    Fetching license for google/ml-metadata
+    Fetching license for kubernetes-sigs/controller-runtime
+    Fetching license for kubernetes-sigs/testing_frameworks
+    Fetching license for kubernetes-sigs/yaml
+    Fetched github license info, 91 succeeded, 0 failed.
+    ```
+    You have to create a `<github_token_file>` in order to access Github repositories. You shall see `license_info.csv` file produced.
+
+2. Update license types.
+
+
+    If you open `license_info.csv`, you can see some fields are marked as `Other`. We have to update them to the right license types. First we need to grep all these unknown license URLs:
+    ```
+    $ cat license_info.csv | grep Other | cut -d ',' -f 2
+
+    GoogleCloudPlatform/gcloud-golang,https://github.com/googleapis/google-cloud-go/blob/master/LICENSE,Other,https://raw.githubusercontent.com/googleapis/google-cloud-go/master/LICENSE
+    ghodss/yaml,https://github.com/ghodss/yaml/blob/master/LICENSE,Other,https://raw.githubusercontent.com/ghodss/yaml/master/LICENSE
+    gogo/protobuf,https://github.com/gogo/protobuf/blob/master/LICENSE,Other,https://raw.githubusercontent.com/gogo/protobuf/master/LICENSE
+    ......
+    ```
+
+    Now we can open these license all at once in Chrome via a plugin called [OpenList](https://chrome.google.com/webstore/detail/openlist/nkpjembldfckmdchbdiclhfedcngbgnl?hl=en).
+
+    After checking the license content one by one, we can now create `additional_license_info.csv` to record the right license types. The content  of `additional_license_info.csv` looks like this:
+    ```
+    https://github.com/googleapis/google-cloud-go/blob/master/LICENSE,Apache License 2.0
+    https://github.com/ghodss/yaml/blob/master/LICENSE,MIT
+    https://github.com/gogo/protobuf/blob/master/LICENSE,BSD 3-Clause "New" or "Revised" License
+    ......
+    ```
+
+    Finally, we can patch the additional license types in `additional_license_info.csv` on `license_info.csv` to get the final list of licenses with types.
+
+    ```
+    $  python patch_additional_license_info.py
+    ```
```

Comment:
Thank you! This is awesome!
Workflow for these is exactly something I didn't have time to polish.

Can I copy this script back to kfp repo? (I'm not sure if there would be legal issues.)

I think it would be the easiest for up-comers to find all relevant tools in one place. What extra efforts do we need to take if we all move these into a dedicated repo? (asking just because I'm not familiar with usual process)

Reply:
@Bobgy and @zhenghuiwang 
We use kubeflow/testing as our engprod repository. If we are building tooling that can be reused across applications for satisfying licensing requirements (I hope we are). Then I would suggest putting that tooling in that repository.

@Bobgy does that work for you?

Reply:
Moving all scripts of image building process into one place like kubeflow/testing SGTM.

@Bobgy If you agree, you could copy your cli to a folder under kubeflow/testing and I can update it.

Reply:
sounds good to me, I will make a PR